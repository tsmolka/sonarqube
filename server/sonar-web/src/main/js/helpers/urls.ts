/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { stringify } from 'querystring';
import { omitBy, isNil } from 'lodash';
import { isShortLivingBranch } from './branches';
import { getProfilePath } from '../apps/quality-profiles/utils';
import { Branch, HomePage, HomePageType } from '../app/types';

interface Query {
  [x: string]: string | undefined;
}

interface Location {
  pathname: string;
  query?: Query;
}

export function getBaseUrl(): string {
  return (window as any).baseUrl;
}

export function getHostUrl(): string {
  return window.location.origin + getBaseUrl();
}

export function getPathUrlAsString(path: Location): string {
  return `${getBaseUrl()}${path.pathname}?${stringify(omitBy(path.query, isNil))}`;
}

export function getProjectUrl(key: string, branch?: string): Location {
  return { pathname: '/dashboard', query: { id: key, branch } };
}

export function getPortfolioUrl(key: string): Location {
  return { pathname: '/portfolio', query: { id: key } };
}

export function getComponentBackgroundTaskUrl(componentKey: string, status?: string): Location {
  return { pathname: '/project/background_tasks', query: { id: componentKey, status } };
}

export function getProjectBranchUrl(key: string, branch: Branch): Location {
  if (isShortLivingBranch(branch)) {
    return {
      pathname: '/project/issues',
      query: { branch: branch.name, id: key, resolved: 'false' }
    };
  } else if (!branch.isMain) {
    return { pathname: '/dashboard', query: { branch: branch.name, id: key } };
  } else {
    return { pathname: '/dashboard', query: { id: key } };
  }
}

/**
 * Generate URL for a global issues page
 */
export function getIssuesUrl(query: Query, organization?: string): Location {
  const pathname = organization ? `/organizations/${organization}/issues` : '/issues';
  return { pathname, query };
}

/**
 * Generate URL for a component's issues page
 */
export function getComponentIssuesUrl(componentKey: string, query?: Query): Location {
  return { pathname: '/project/issues', query: { ...(query || {}), id: componentKey } };
}

/**
 * Generate URL for a component's drilldown page
 */
export function getComponentDrilldownUrl(
  componentKey: string,
  metric: string,
  branch?: string
): Location {
  return { pathname: '/component_measures', query: { id: componentKey, metric, branch } };
}

export function getMeasureTreemapUrl(component: string, metric: string, branch?: string) {
  return {
    pathname: '/component_measures',
    query: { id: component, metric, branch, view: 'treemap' }
  };
}

/**
 * Generate URL for a component's measure history
 */
export function getMeasureHistoryUrl(component: string, metric: string, branch?: string) {
  return {
    pathname: '/project/activity',
    query: { id: component, graph: 'custom', custom_metrics: metric, branch }
  };
}

/**
 * Generate URL for a component's permissions page
 */
export function getComponentPermissionsUrl(componentKey: string): Location {
  return { pathname: '/project_roles', query: { id: componentKey } };
}

/**
 * Generate URL for a quality profile
 */
export function getQualityProfileUrl(
  name: string,
  language: string,
  organization?: string | null
): Location {
  return getProfilePath(name, language, organization);
}

export function getQualityGateUrl(key: string, organization?: string | null): Location {
  return {
    pathname: getQualityGatesUrl(organization).pathname + '/show/' + encodeURIComponent(key)
  };
}

export function getQualityGatesUrl(organization?: string | null): Location {
  return {
    pathname:
      (organization ? '/organizations/' + encodeURIComponent(organization) : '') + '/quality_gates'
  };
}

/**
 * Generate URL for the rules page
 */
export function getRulesUrl(query: Query, organization: string | null | undefined): Location {
  const pathname = organization ? `/organizations/${organization}/rules` : '/coding_rules';
  return { pathname, query };
}

/**
 * Generate URL for the rules page filtering only active deprecated rules
 */
export function getDeprecatedActiveRulesUrl(
  query: Query = {},
  organization: string | null | undefined
): Location {
  const baseQuery = { activation: 'true', statuses: 'DEPRECATED' };
  return getRulesUrl({ ...query, ...baseQuery }, organization);
}

export function getRuleUrl(rule: string, organization: string | undefined) {
  /* eslint-disable camelcase */
  return getRulesUrl({ open: rule, rule_key: rule }, organization);
  /* eslint-enable camelcase */
}

export function getMarkdownHelpUrl(): string {
  return getBaseUrl() + '/markdown/help';
}

export function getCodeUrl(project: string, branch?: string, selected?: string) {
  return { pathname: '/code', query: { id: project, branch, selected } };
}

export function getOrganizationUrl(organization: string) {
  return `/organizations/${organization}`;
}

export function getHomePageUrl(homepage: HomePage) {
  switch (homepage.type) {
    case HomePageType.Application:
      return getProjectUrl(homepage.component);
    case HomePageType.Project:
      return getProjectUrl(homepage.component, homepage.branch);
    case HomePageType.Organization:
      return getOrganizationUrl(homepage.organization);
    case HomePageType.Portfolio:
      return getPortfolioUrl(homepage.component);
    case HomePageType.Portfolios:
      return '/portfolios';
    case HomePageType.MyProjects:
      return '/projects';
    case HomePageType.Issues:
    case HomePageType.MyIssues:
      return { pathname: '/issues', query: { resolved: 'false' } };
  }

  // should never happen, but just in case...
  return '/projects';
}

export function isUrl(url: string) {
  if (!URL) {
    const elem = document.createElement('a');
    elem.href = url;
    return !!(elem.host && elem.hostname && elem.protocol);
  }
  try {
    const parsedUrl = new URL(url);
    return url.includes(parsedUrl.host);
  } catch (error) {
    return false;
  }
}
