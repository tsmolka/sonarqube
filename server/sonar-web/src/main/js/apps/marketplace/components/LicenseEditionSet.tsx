/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { stringify } from 'querystring';
import * as React from 'react';
import * as classNames from 'classnames';
import { FormattedMessage } from 'react-intl';
import { debounce } from 'lodash';
import Checkbox from '../../../components/controls/Checkbox';
import DeferredSpinner from '../../../components/common/DeferredSpinner';
import { omitNil } from '../../../helpers/request';
import { Edition, getFormData, getLicensePreview } from '../../../api/marketplace';
import { translate, translateWithParameters } from '../../../helpers/l10n';

export interface Props {
  className?: string;
  edition?: Edition;
  editions: Edition[];
  updateLicense: (license?: string, status?: string) => void;
}

interface State {
  acceptTerms: boolean;
  formData?: {
    serverId?: string;
    ncloc?: number;
  };
  license: string;
  licenseEdition?: Edition;
  loading: boolean;
  previewStatus?: string;
  wrongEdition: boolean;
}

export default class LicenseEditionSet extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);
    this.state = { acceptTerms: false, license: '', loading: false, wrongEdition: false };
    this.fetchLicensePreview = debounce(this.fetchLicensePreview, 100);
  }

  componentDidMount() {
    this.mounted = true;
    this.fetchFormData();
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchLicensePreview = (license: string) => {
    this.setState({ loading: true });
    getLicensePreview({ license }).then(
      ({ previewStatus, nextEditionKey }) => {
        if (this.mounted) {
          const { edition } = this.props;
          const licenseEdition = this.props.editions.find(
            edition => edition.key === nextEditionKey
          );
          const wrongEdition = Boolean(
            !licenseEdition || (edition && edition.key !== nextEditionKey)
          );
          this.setLicense({ license, loading: false, licenseEdition, previewStatus, wrongEdition });
        }
      },
      () => {
        if (this.mounted) {
          this.resetLicense({ license, loading: false });
        }
      }
    );
  };

  fetchFormData = () => {
    getFormData().then(
      formData => {
        if (this.mounted) {
          this.setState({ formData });
        }
      },
      () => {}
    );
  };

  getLicenseFormUrl = (edition: Edition) => {
    let url = edition.licenseRequestUrl;
    if (this.state.formData) {
      const query = stringify(omitNil(this.state.formData));
      if (query) {
        url += '?' + query;
      }
    }
    return url;
  };

  handleLicenseChange = (event: React.SyntheticEvent<HTMLTextAreaElement>) => {
    const license = event.currentTarget.value;
    if (license) {
      this.fetchLicensePreview(license);
      this.setState({ license });
    } else {
      this.resetLicense();
    }
  };

  handleTermsCheck = (checked: boolean) => {
    this.setLicense({ acceptTerms: checked });
  };

  resetLicense = (state = {}) => {
    this.setLicense({
      license: '',
      licenseEdition: undefined,
      previewStatus: undefined,
      wrongEdition: false,
      ...state
    });
  };

  setLicense = (state = {}) => {
    this.setState(state, this.updateParentLicense);
  };

  updateParentLicense = () => {
    const { acceptTerms, license, previewStatus, wrongEdition } = this.state;
    this.props.updateLicense(
      previewStatus !== 'NO_INSTALL' && !acceptTerms ? undefined : license,
      wrongEdition ? undefined : previewStatus
    );
  };

  renderAlert() {
    const { licenseEdition, previewStatus, wrongEdition } = this.state;
    if (!previewStatus || wrongEdition) {
      const { edition } = this.props;

      return (
        <div className="spacer-top">
          {wrongEdition && (
            <p className="alert alert-danger">
              {edition
                ? translateWithParameters('marketplace.wrong_license_type_x', edition.name)
                : translate('marketplace.wrong_license_type')}
            </p>
          )}
          {edition && (
            <a href={this.getLicenseFormUrl(edition)} target="_blank">
              {translate('marketplace.i_need_a_license')}
            </a>
          )}
        </div>
      );
    }

    return (
      <div className="spacer-top">
        <p
          className={classNames('alert', {
            'alert-warning': previewStatus === 'AUTOMATIC_INSTALL',
            'alert-success': previewStatus === 'NO_INSTALL',
            'alert-danger': previewStatus === 'MANUAL_INSTALL'
          })}>
          {translateWithParameters(
            'marketplace.license_preview_status.' + previewStatus,
            licenseEdition ? licenseEdition.name : translate('marketplace.commercial_edition')
          )}
          {licenseEdition &&
            licenseEdition.key === 'datacenter' &&
            previewStatus !== 'NO_INSTALL' && (
              <span className="little-spacer-left">
                <FormattedMessage
                  defaultMessage={translate('marketplace.how_to_setup_cluster_url')}
                  id="marketplace.how_to_setup_cluster_url"
                  values={{
                    url: (
                      <a
                        href="https://redirect.sonarsource.com/doc/data-center-edition.html"
                        rel="noopener noreferrer"
                        target="_blank">
                        {licenseEdition.name}
                      </a>
                    )
                  }}
                />
              </span>
            )}
        </p>
        {previewStatus !== 'NO_INSTALL' && (
          <span className="js-edition-tos">
            <Checkbox
              checked={this.state.acceptTerms}
              id="edition-terms"
              onCheck={this.handleTermsCheck}>
              <label className="little-spacer-left" htmlFor="edition-terms">
                {translate('marketplace.i_accept_the')}
              </label>
            </Checkbox>
            <a
              className="nowrap little-spacer-left"
              href="http://dist.sonarsource.com/SonarSource_Terms_And_Conditions.pdf"
              rel="noopener noreferrer"
              target="_blank">
              {translate('marketplace.terms_and_conditions')}
            </a>
          </span>
        )}
      </div>
    );
  }

  render() {
    const { className, edition } = this.props;
    const { license, loading } = this.state;

    return (
      <div className={className}>
        {edition && (
          <label className="display-inline-block spacer-bottom" htmlFor="set-license">
            {translateWithParameters('marketplace.enter_license_for_x', edition.name)}
            <em className="mandatory">*</em>
          </label>
        )}
        <textarea
          autoFocus={true}
          className="display-block input-super-large"
          id="set-license"
          onChange={this.handleLicenseChange}
          required={true}
          rows={8}
          style={{ resize: 'none' }}
          value={license}
        />

        <DeferredSpinner
          customSpinner={
            <p className="spacer-top">
              <i className="spinner spacer-right text-bottom" />
              {translate('marketplace.checking_license')}
            </p>
          }
          loading={loading}>
          {this.renderAlert()}
        </DeferredSpinner>
      </div>
    );
  }
}
