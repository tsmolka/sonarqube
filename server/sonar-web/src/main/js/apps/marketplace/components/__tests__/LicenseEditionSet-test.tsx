/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable import/order */
import * as React from 'react';
import { shallow } from 'enzyme';
import { change, waitAndUpdate } from '../../../../helpers/testUtils';
import LicenseEditionSet from '../LicenseEditionSet';

jest.mock('../../../../api/marketplace', () => ({
  getLicensePreview: jest.fn(() =>
    Promise.resolve({ nextEditionKey: 'foo', previewStatus: 'NO_INSTALL' })
  ),
  getFormData: jest.fn(() => Promise.resolve({ serverId: 'foo', ncloc: 1000 }))
}));

jest.mock('lodash', () => {
  const lodash = require.requireActual('lodash');
  lodash.debounce = (fn: Function) => (...args: any[]) => fn(...args);
  return lodash;
});

const getLicensePreview = require('../../../../api/marketplace').getLicensePreview as jest.Mock<
  any
>;

const DEFAULT_EDITION = {
  key: 'foo',
  name: 'Foo',
  textDescription: 'Foo desc',
  downloadUrl: 'download_url',
  homeUrl: 'more_url',
  licenseRequestUrl: 'license_url'
};

beforeEach(() => {
  getLicensePreview.mockClear();
});

it('should display correctly', () => {
  expect(getWrapper()).toMatchSnapshot();
});

it('should display the get license link with parameters', async () => {
  const wrapper = getWrapper();
  await waitAndUpdate(wrapper);
  expect(wrapper.find('a')).toMatchSnapshot();
});

it('should correctly display status message after checking license', async () => {
  let wrapper = await testLicenseStatus('NO_INSTALL');
  expect(wrapper.find('p.alert')).toMatchSnapshot();
  wrapper = await testLicenseStatus('AUTOMATIC_INSTALL');
  expect(wrapper.find('p.alert')).toMatchSnapshot();
  wrapper = await testLicenseStatus('MANUAL_INSTALL');
  expect(wrapper.find('p.alert')).toMatchSnapshot();
  wrapper = await testLicenseStatus('AUTOMATIC_INSTALL', jest.fn(), 'bar');
  expect(wrapper.find('p.alert')).toMatchSnapshot();
  wrapper = await testLicenseStatus('AUTOMATIC_INSTALL', jest.fn(), 'bar', { edition: undefined });
  expect(wrapper.find('p.alert')).toMatchSnapshot();
});

it('should display terms of license checkbox', async () => {
  let updateLicense = jest.fn();
  let wrapper = await testLicenseStatus('NO_INSTALL', updateLicense);
  expect(wrapper.find('.js-edition-tos').exists()).toBeFalsy();
  expect(updateLicense).toHaveBeenCalledWith('mylicense', 'NO_INSTALL');

  updateLicense = jest.fn();
  wrapper = await testLicenseStatus('AUTOMATIC_INSTALL', updateLicense);
  const tosCheckbox = wrapper.find('.js-edition-tos');
  expect(tosCheckbox.find('a').exists()).toBeTruthy();
  expect(updateLicense).toHaveBeenLastCalledWith(undefined, 'AUTOMATIC_INSTALL');
  (tosCheckbox.find('Checkbox').prop('onCheck') as Function)(true);
  expect(updateLicense).toHaveBeenLastCalledWith('mylicense', 'AUTOMATIC_INSTALL');

  updateLicense = jest.fn();
  wrapper = await testLicenseStatus('MANUAL_INSTALL', updateLicense);
  expect(wrapper.find('.js-edition-tos').exists()).toBeTruthy();
  expect(updateLicense).toHaveBeenLastCalledWith(undefined, 'MANUAL_INSTALL');
});

function getWrapper(props = {}) {
  return shallow(
    <LicenseEditionSet
      edition={DEFAULT_EDITION}
      editions={[DEFAULT_EDITION]}
      updateLicense={jest.fn()}
      {...props}
    />
  );
}

async function testLicenseStatus(
  status: string,
  updateLicense = jest.fn(),
  nextEditionKey = 'foo',
  props = {}
) {
  getLicensePreview.mockImplementation(() =>
    Promise.resolve({ nextEditionKey, previewStatus: status })
  );
  const wrapper = getWrapper({ updateLicense, ...props });
  change(wrapper.find('textarea'), 'mylicense');
  expect(getLicensePreview).toHaveBeenCalled();
  await new Promise(setImmediate);
  expect(updateLicense).toHaveBeenCalled();
  wrapper.update();
  return wrapper;
}
