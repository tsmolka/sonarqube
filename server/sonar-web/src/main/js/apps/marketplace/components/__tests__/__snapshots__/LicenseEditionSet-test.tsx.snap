// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly display status message after checking license 1`] = `
<p
  className="alert alert-success"
>
  marketplace.license_preview_status.NO_INSTALL.Foo
</p>
`;

exports[`should correctly display status message after checking license 2`] = `
<p
  className="alert alert-warning"
>
  marketplace.license_preview_status.AUTOMATIC_INSTALL.Foo
</p>
`;

exports[`should correctly display status message after checking license 3`] = `
<p
  className="alert alert-danger"
>
  marketplace.license_preview_status.MANUAL_INSTALL.Foo
</p>
`;

exports[`should correctly display status message after checking license 4`] = `
<p
  className="alert alert-danger"
>
  marketplace.wrong_license_type_x.Foo
</p>
`;

exports[`should correctly display status message after checking license 5`] = `
<p
  className="alert alert-danger"
>
  marketplace.wrong_license_type
</p>
`;

exports[`should display correctly 1`] = `
<div>
  <label
    className="display-inline-block spacer-bottom"
    htmlFor="set-license"
  >
    marketplace.enter_license_for_x.Foo
    <em
      className="mandatory"
    >
      *
    </em>
  </label>
  <textarea
    autoFocus={true}
    className="display-block input-super-large"
    id="set-license"
    onChange={[Function]}
    required={true}
    rows={8}
    style={
      Object {
        "resize": "none",
      }
    }
    value=""
  />
  <DeferredSpinner
    customSpinner={
      <p
        className="spacer-top"
      >
        <i
          className="spinner spacer-right text-bottom"
        />
        marketplace.checking_license
      </p>
    }
    loading={false}
    timeout={100}
  >
    <div
      className="spacer-top"
    >
      <a
        href="license_url"
        target="_blank"
      >
        marketplace.i_need_a_license
      </a>
    </div>
  </DeferredSpinner>
</div>
`;

exports[`should display the get license link with parameters 1`] = `
<a
  href="license_url?serverId=foo&ncloc=1000"
  target="_blank"
>
  marketplace.i_need_a_license
</a>
`;
